lib/fs.js

```diff
@@ -2479,3 +2479,454 @@ Object.defineProperty(fs, 'SyncWriteStream', {
+                 pathModule.toNamespacedPath(newPath),
+                 req);
+    return req.promise;
+  },
+
+  async lstat(path) {
+    handleError((path = getPathFromURL(path)));
+    nullCheck(path);
+    const req = new FSReqPromise();
+    binding.lstat(pathModule.toNamespacedPath(path), req);
+    return statsFromValues(await req.promise);
+  },
+
+  async mkdir(path, mode) {
+    handleError((path = getPathFromURL(path)));
+    nullCheck(path);
+    const req = new FSReqPromise();
+    binding.mkdir(pathModule.toNamespacedPath(path),
+                  modeNum(mode, 0o777),
+                  req);
+    return req.promise;
+  },
+
+  async mkdtemp(prefix, options) {
+    options = getOptions(options, {});
+    if (!prefix || typeof prefix !== 'string') {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
+                                 'prefix',
+                                 'string',
+                                 prefix);
+    }
+    nullCheck(prefix);
+    const req = new FSReqPromise();
+    binding.mkdtemp(`${prefix}XXXXXX`, options.encoding, req);
+    return req.promise;
+  },
+
+  async open(path, flags, mode) {
+    mode = modeNum(mode, 0o666);
+    handleError((path = getPathFromURL(path)));
+    nullCheck(path);
+    const req = new FSReqPromise();
+    binding.openFD(pathModule.toNamespacedPath(path),
+                   stringToFlags(flags),
+                   mode,
+                   req);
+    return req.promise;
+  },
+
+  async read(fd, buffer, offset, length, position) {
+    if (fd == null || typeof fd !== 'object' || !Number.isInteger(fd.fd))
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'fd', 'FD');
+    const fdnum = fd.fd;
+    if (fdnum < 0 || fdnum > 0xFFFFFFFF)
+      throw new errors.RangeError('ERR_OUT_OF_RANGE', 'fd');
+    if (!isUint8Array(buffer))
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'buffer',
+                                 ['Buffer', 'Uint8Array']);
+
+    offset |= 0;
+    length |= 0;
+
+    if (length === 0)
+      return { length: 0, buffer };
+
+    if (offset < 0 || offset >= buffer.length)
+      throw new errors.RangeError('ERR_OUT_OF_RANGE', 'offset');
+
+    if (length < 0 || offset + length > buffer.length)
+      throw new errors.RangeError('ERR_OUT_OF_RANGE', 'length');
+
+    if (!Number.isInteger(position))
+      position = -1;
+
+    const req = new FSReqPromise();
+    binding.read(fdnum, buffer, offset, length, position, req);
+    const bytesRead = await req.promise;
+    return { length: bytesRead, buffer };
+  },
+
+  async readFile(path, options) {
+    options = getOptions(options, { flag: 'r' });
+
+    if (typeof path === 'object' && Number.isInteger(path.fd))
+      return readFd(path);
+
+    return readFd(await fs.promises.open(path, options.flag, 0o666));
+
+    async function readFd(fd) {
+      let ret = Buffer.alloc(0);
+      for await(const chunk of readEverything(fd))
+        ret = Buffer.concat([ret, chunk]);
```

Comment:
I'm afraid it's bad for performance to concat for each chunk. Is it possible to push to an array and concat once we are out of the loop?

Reply:
yep, whatever works best.

Reply:
for await is very slow at the moment

Reply:
yep... part of my goal is to make this available then see what @bmeurer and team can do in terms of making it faster.

Reply:
@jasnell this would make the promise API prohibitebly expensive to use though.

Reply:
While the API is still experimental, performance is not the priority. If we're not able to get the performance optimized adequately, then we can move to a different approach.